src/net.cpp

```diff
@@ -496,8 +496,21 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
             if (m_i2p_sam_session) {
                 connected = m_i2p_sam_session->Connect(addrConnect, conn, proxyConnectionFailed);
             } else {
-                i2p_transient_session = std::make_unique<i2p::sam::Session>(proxy.proxy, &interruptNet);
+                {
+                    LOCK(m_i2p_sessions_mutex);
+                    if (m_i2p_sessions.empty()) {
+                        i2p_transient_session =
+                            std::make_unique<i2p::sam::Session>(proxy.proxy, &interruptNet);
+                    } else {
+                        i2p_transient_session.swap(m_i2p_sessions.front());
+                        m_i2p_sessions.pop();
+                    }
+                }
                 connected = i2p_transient_session->Connect(addrConnect, conn, proxyConnectionFailed);
+                if (!connected) {
+                    LOCK(m_i2p_sessions_mutex);
+                    m_i2p_sessions.emplace(i2p_transient_session.release());
+                }
```

Comment:
I am wondering if `m_i2p_sessions` can grow unchecked to the point of using too much memory and whether it should be capped.

If this is run by a single thread then the size of `m_i2p_sessions` can be either 0 or 1 because in order to create a new session, the list must be empty and then if the connection fails it will be added to the list. If run by N threads, then the size can be at most N, if all of the threads race into it.

I think that is ok with the current pattern of add 1, take 1. Just a note to the future - if the code is changed to add more entries to the list and it takes.

Reply:
Maybe only emplace if under a max size.

```diff
-                if (!connected) {
+                if (!connected && m_i2p_sessions.size() < 10) {
                     LOCK(m_i2p_sessions_mutex);
                     m_i2p_sessions.emplace(i2p_transient_session.release());
```

Reply:
Fair enough, not too difficult to add a cap. Done.